README.md
# MD.js 

- ~1kb minified+gzipped
- ~1.5kb minified

[![npm](https://img.shields.io/npm/v/md.js.svg?style=flat)](https://www.npmjs.com/package/md.js) [![licence](https://img.shields.io/badge/licence-MIT-blue.svg?style=flat)](https://github.com/thysultan/md.js/blob/master/LICENSE.md)

MD is a lightweight markdown parser with default support for langauge specific code blocks and XSS filtering, for example

```javascript
// replace . with `
```javascript
	var int = 0;
```
```

will output the following

```html
<pre><code class="language-javascript">
	var int = 0;
</code></pre>
```

XSS attempt `<script>alert(1);</script>` will be converted to `&lt;script&gt;alert(1)&lt;/script&gt;`
while `<a href="javascript:alert('xss')"></a>` will be converted to `<a href="#javascript&#58;alert('xss')"></a>`
and `<img onerror="alert(1)">` inline events will be remove and more.

## Browser Support

* Edge
* IE 8+
* Chrome
* Firefox
* Safari

# Installation

#### direct download

```html
<script src=md.min.js></script>
```

#### CDN

```html
<script src=https://unpkg.com/md.js@0.2.6/md.min.js></script>
```

#### npm

```
npm install md.js --save
```

#### Useage

```javascript
md('# Heading...');
```

# Examples

[demo](https://rawgit.com/thysultan/md.js/master/examples/index.html)

CHANGELOG.md
## 0.2.6 (January 10, 2016)

- patch blockquote token `>` in inline code `() => {}`

tests/index.js
/**
 * tester runner
 *
 * excutes tests given an object describing the tests
 * based on sample input vs expected output
 * if output === expected output, test passed
 * else test failed
 *
 * log format
 *
 * ------
 * 
 * Tests Passed #
 *
 * ...
 *
 * [Finnished In] #ms
 * 
 * Tests Failed #
 *
 * ...
 *
 * [Finnished In] #ms
 */


var browser = this && !!this.window;
var md = browser ? this.md : require('../md.js');

/**
 * run tests
 * @return {Object} tests
 */
function run (tests) {
	var start = Date.now();

	var passed = [];
	var failed = [];

	var format = {
		reset:     browser ? '' : '\x1b[0m',
		green:     browser ? '' : '\x1b[32m',
		red:       browser ? '' : '\x1b[31m',
		yellow:    browser ? '' : '\x1b[33m',
		underline: browser ? '' : '\x1b[4m',
		dim:       browser ? '' : '\x1b[2m',
		bold:      browser ? '' : '\x1b[1m',
		clear:     browser ? '' : '\x1Bc\n'
	};

	for (var name in tests) {
		var test = tests[name];

		var name = test.name.trim();
		var sample = test.sample.trim();
		var expected = test.expected.trim();

		var result = md(sample);

		(result === expected ? passed : failed).push(name);

		if (result !== expected) {
			// log why it failed
			console.log('failed:\n'+ result);
			console.log('expected:\n'+ expected);
		}
	}

	var end = '\n\n'+format.reset+'[Finnished In] '+(Date.now()-start)+'ms\n';

	// start test logger
	console.log('\n------');

	// passed
	console.log(
		format.bold+'\nTests Passed '+passed.length+format.reset+format.green + '\n\n'+passed.join('\n')+end
	);

	// failed
	console.log(
		format.bold+'Tests Failed '+failed.length+format.reset+format.red + '\n\n'+(failed.join('\n') || 'no failed tests')+end
	);

	// if failed trigger exit
	if (failed.length) {
		if (browser) {
			console.error(new Error('^^^'));
		} else {
			process.exit(1);
		}
	}
}


/**
 * define tests
 * @type {Object}
 */
var tests = {
	headings: {
		name: 'headings',
		sample: `
			# Heading 1
			## Heading 2
			### Heading 3
			#### Heading 4
			Heading 1
			====
			Heading 2
			----
		`,
		expected: `
<h1>Heading 1</h1><h2>Heading 2</h2>
<h3>Heading 3</h3><h4>Heading 4</h4>
<h1>Heading 1</h1><h2>Heading 2</h2>
		`
	},
	lists: {
		name: 'lists',
		sample: `
			- [x] ul list 1
			- [ ] ul list 2

			- ul list 1
			- ul list 2

			* ul list 1
			* ul list 2
			* ul list 3

			+ ul list 1
			+ ul list 2

			1. ol list 1
			2. ol list 2
			3. ol list 3

			1) ol list 1
			2) ol list 2
			3) ol list 3
		`,
		expected: `
<ul><li><input type="checkbox" disabled checked> ul list 1</li><li><input type="checkbox" disabled> ul list 2</li></ul>

<ul><li>ul list 1</li><li>ul list 2</li></ul>

<ul><li>ul list 1</li><li>ul list 2</li><li>ul list 3</li></ul>

<ul><li>ul list 1</li><li>ul list 2</li></ul>

<ol><li>ol list 1</li><li>ol list 2</li><li>ol list 3</li></ol>

<ol><li>ol list 1</li><li>ol list 2</li><li>ol list 3</li></ol>
		`
	},
	blockquotes: {
		name: 'blockquotes',
		sample: `> hello world what is this`,
		expected: `<blockquote>hello world what is this</blockquote>`
	},
	horizontal: {
		name: 'horizontal rule',
		sample: `
		---
		***
		- - -
		* * *
		`,
		expected: `
<hr><hr>
<hr><hr>
		`
	},
	images: {
		name: 'images',
		sample: `![Alt text](https://thysultan.com/dio/assets/logo.svg)`,
		expected: `<img src="https://thysultan.com/dio/assets/logo.svg" alt="Alt text">`
	},
	links: {
		name: 'links',
		sample: `[an example](http://example.com/ "Title")`,
		expected: `<a href="http://example.com/" title="Title">an example</a>`
	},
	paragraphs: {
		name: 'paragraphs',
		sample: `
			Lorem ipsum dolor sit amet, _consectetur_ adipisicing elit. 
			Cupiditate doloremque sed vero __excepturi__.
			Lorem ipsum dolor sit amet, *consectetur* adipisicing elit. 
			Cupiditate doloremque sed vero **excepturi**.
			Another paragraphs # Heading, will not match this
		`,
		expected: `
<p>Lorem ipsum dolor sit amet, <em>consectetur</em> adipisicing elit.
Cupiditate doloremque sed vero <strong>excepturi</strong>.
Lorem ipsum dolor sit amet, <em>consectetur</em> adipisicing elit.
Cupiditate doloremque sed vero <strong>excepturi</strong>.
Another paragraphs # Heading, will not match this</p>
		`
	},
	emphasis: {
		name: 'emphasis',
		sample: `
			_consectetur_
			*excepturi*
		`,
		expected: `
<em>consectetur</em>
<em>excepturi</em>
		`
	},
	strong: {
		name: 'strong',
		sample: `
			__consectetur__
			**excepturi**
		`,
		expected: `
<strong>consectetur</strong>
<strong>excepturi</strong>
		`
	},
	strike: {
		name: 'strike',
		sample: `~~consectetur~~`,
		expected: `<p><del>consectetur</del></p>`
	},
	html: {
		name: 'html',
		sample: `
			<div style="color:red">This is a styled html div</div>
		`,
		expected: `
<div style="color:red">This is a styled html div</div>
		`
	},
	code: {
		name: 'code',
		sample: 'Use the `() => {}` function\n\n```javascript\n\t\tvar i = 10;\n\tvar foo = 2;<script>1</script>\n```',
		expected: '<p>Use the <code>&#40;&#41; =&gt; {}</code> function</p>\n'+
		'<pre><code class="language-javascript">\t\tvar i = 10;\n\tvar foo = 2;&lt;script&gt;1&lt;/script&gt;\n</code></pre>'
	}
};


/**
 * execute tests
 */
run(tests);

examples/example.js
var markdown = (
	'Use the `printf()` function.\n\n'+

	'```javascript\n\n\t\tvar i = 10;\nvar foo = 2;\n<script>1</script>\n```' +

`
	# Heading 1
	## Heading 2
	### Heading 3
	#### *Heading 4*

	<img onerror="alert()">
	<img src="" alt="<a href="onerror='alert(String.fromCharCode(88,83,83))'b="></a>">

	Heading 1 (commonmark)
	====

	Heading 2 (commonmark)
	----

	![Alt text](https://thysultan.com/dio/assets/logo.svg)

	This is [an example](http://example.com/ "Title") inline link.

	***

	Paragraphs Lorem ipsum ~~strike this~~ sit amet, _consectetur_ adipisicing elit. 
	Cupiditate doloremque sed vero __excepturi__.  

	Lorem ipsum dolor sit amet, *consectetur* adipisicing elit. 
	Cupiditate doloremque sed vero **excepturi**.

	[a link](http://google.com)

	- [ ] ul list 1
	- [x] ul list 2

	- ul list 1
	- ul list 2

	* ul list 1
	* ul list 2
	* ul list 3

	+ ul list 1
	+ ul list 2

	1. ol list 1
	2. ol list 2
	3. ol list 3

	1) ol list 1
	2) ol list 2
	3) ol list 3

	> hello world what is this

	---

	<table>
	    <tr>
	        <td>This is an html table block</td>
	    </tr>
	</table>

	<div style="color:red">This is a styled html div</div>

	<div>
		Isolated Text In html block
	</div>

	<style>
		#foo {
			color: red;
		}
	</style>

	XSS With Image ![]("onerror="alert())

	XSS Anchors [test]("onerror="alert())

	Another paragraphs # Heading, will not match this
	Escape \\* should world and so should \\# as well as \\~~strike right?

	## XSS Tests

	<IMG SRC="javascript:alert('XSS');">

	
	<a href="javascript:alert('xss')">*you*</a>

	
	XSS filter <script>alert(1);</script> should not alert.

	
	XSS filter upperCase <SCRIPT>alert(1);</SCRIPT> should not alert.


	XSS filter href <a href="javascript:alert('xss')">you</a>


	XSS filter upperCase href <a HREF="javascript:alert('xss')">you</a>


	<IMG SRC= onmouseover="alert('xxs')">


	<IMG """><SCRIPT>alert("XSS")</SCRIPT>">


	<IMG SRC="jav	ascript:alert('XSS');">


	<IMG SRC="jav&#x0A;ascript:alert('XSS');">

	---

	Trailing Paragraph Lorem ipsum dolor sit amet, _consectetur_ adipisicing elit. 
	Cupiditate doloremque sed vero __excepturi__.`);

// markdown = '#  Title\n\nAnd *now* [a link](http://www.google.com) to **follow** and [another](http://yahoo.com/).\n\n* One\n* Two\n* Three\n\n## Subhead\n\nOne **two** three **four** five.\n\nOne __two__ three _four_ five __six__ seven _eight_.\n\n1. One\n2. Two\n3. Three\n\nMore text with `inline($code)` and :"quote": sample.\n\n> A block quote\n> across two lines.\nMore text...';

var html = md(markdown);
document.body.insertAdjacentHTML('beforeend', html);

// console.log(html);

console.log(md('![](" onerror="alert())'));

console.log(md('<img src="" onerror=onerror="alert()">'));

console.log(md('<script language="text/javascript">alert()</script>'));

console.log(md("![[](onerror='alert(String.fromCharCode(88,83,83))'b=)]()"));

console.log(md(`
<blockquote>
	<p>hello <a name="n"
	href="javascript:alert('xss')"><em>you</em></a></p>
</blockquote>
`));

console.log(md(`
	[some text](javascript:alert('xss'))
`));

md.js
/*!
 *             __
 *   __ _  ___/ /
 *  /  ' \/ _  / 
 * /_/_/_/\_,_/ 
 * 
 * md.js is a lightweight markdown parser
 * https://github.com/thysultan/md.js
 * 
 * @licence MIT
 */
(function (factory) {
	if (typeof exports === 'object' && typeof module !== 'undefined') {
		module.exports = factory();
	} else if (typeof define === 'function' && define.amd) {
		define(factory());
	} else {
		window.md = factory();
	}
}(function () {
	var unicodes = {
		'<': '&lt;',
		'>': '&gt;',
		'"': '&quot;',
		"'": '&#39;',
		'&': '&amp;',
		'[': '&#91;',
		']': '&#93;',
		'(': '&#40;',
		')': '&#41;',
	};

	var resc = /[<>&\(\)\[\]"']/g;

	function unicode (char) { return unicodes[char] || char; }

	var XSSFilterRegExp = /<(script)[^\0]*?>([^\0]+?)<\/(script)>/gmi;
	var XSSFilterTemplate = '&lt;$1&gt;$2&lt;/$3&gt;';

	var XSSFilterInlineJSRegExp = /(<.*? [^\0]*?=[^\0]*?)(javascript:.*?)(.*>)/gmi;
	var XSSFilterInlineJSTemplate = '$1#$2&#58;$3';

	var XSSFilterImageRegExp = /<img([^\0]*?onerror=)([^\0]*?)>/gmi;
	var XSSFilterImageTemplate = function (match, group1, group2) {
		return '<img' + group1 + group2.replace(resc, unicode) + '>';
	};

	var removeTabsRegExp = /^[\t ]+|[\t ]$/gm;

	var htmlFilterRegExp = /(<.*>[\t ]*\n^.*)/gm;
	var htmlFilterTemplate = function (match, group1) { 
		return group1.replace(/^\n|$\n/gm, '');
	};

	var cssFilterRegExp = /(<style>[^]*<\/style>)/gm;
	var cssFilterTemplate = htmlFilterTemplate;

	var eventsFilterRegExp = /(<[^]+?)(on.*?=.*?)(.*>)/gm;
	var eventsFilterTemplate = '$1$3';

	var blockQuotesRegExp = /^[ \t]*> (.*)/gm;
	var blockQuotesTemplate = '<blockquote>$1</blockquote>';

	var inlineCodeRegExp = /`([^`]+?)`/g;
	var inlineCodeTemplate = function (match, group1) {
		return '<code>'+group1.replace(resc, unicode)+'</code>'
	}

	var blockCodeRegExp = /```(.*)\n([^\0]+?)```(?!```)/gm;

	var imagesRegExp = /!\[(.*)\]\((.*)\)/g;
	var imagesTemplate = function (match, group1, group2) {
		var src = group2.replace(resc, unicode);
		var alt = group1.replace(resc, unicode);

		return '<img src="'+src+'" alt="'+alt+'">';
	};

	var headingsRegExp = /^(#+) +(.*)/gm;
	var headingsTemplate = function (match, hash, content) {
		var length = hash.length; return '<h'+length+'>'+content+'</h'+length+'>';
	};

	var headingsCommonh2RegExp = /^([^\n\t ])(.*)\n----+/gm;
	var headingsCommonh1RegExp = /^([^\n\t ])(.*)\n====+/gm;
	var headingsCommonh1Template = '<h1>$1$2</h1>';
	var headingsCommonh2Template = '<h2>$1$2</h2>';

	var paragraphsRegExp = /^([^-><#\d\+\_\*\t\n\[\! \{])([^]*?)(|  )(?:\n\n)/gm;
	var paragraphsTemplate = function (match, group1, group2, group3) {
		var leadingCharater = group1;
		var body = group2;
		
		var trailingSpace = group3 ? '\n<br>\n' : '\n';
		return '<p>'+leadingCharater+body+'</p>'+trailingSpace;
	};

	var horizontalRegExp = /^.*?(?:---|\*\*\*|- - -|\* \* \*)/gm;
	var horizontalTemplate = '<hr>';

	var strongRegExp = /(?:\*\*|\_\_)([^\*\n_]+?)(?:\*\*|\_\_)/g;
	var strongTemplate = '<strong>$1</strong>';

	var emphasisRegExp = /(?:\*|\_)([^\*\n_]+?)(?:\*|\_)/g;
	var emphasisTemplate = '<em>$1</em>';

	var strikeRegExp = /(?:~~)([^~]+?)(?:~~)/g;
	var strikeTemplate = '<del>$1</del>';

	var linksRegExp = /\[(.*?)\]\(([^\t\n ]*)(?:| "(.*)")\)+/gm;
	var linksTemplate = function (match, group1, group2, group3) {
		var link = group2.replace(resc, unicode);
		var text = group1.replace(resc, unicode);
		var title = group3 ? ' title="'+group3.replace(resc, unicode)+'"' : '';

		return '<a href="'+link+'"'+title+'>'+text+'</a>';
	};

	var listUlRegExp1 = /^[\t ]*?(?:-|\+|\*) (.*)/gm;
	var listUlRegExp2 = /(\<\/ul\>\n(.*)\<ul\>*)+/g;
	var listUlTemplate = '<ul><li>$1</li></ul>';

	var listOlRegExp1 = /^[\t ]*?(?:\d(?:\)|\.)) (.*)/gm;
	var listOlRegExp2 = /(\<\/ol\>\n(.*)\<ol\>*)+/g;
	var listOlTemplate = '<ol><li>$1</li></ol>';

	var lineBreaksRegExp = /^\n\n+/gm;
	var lineBreaksTemplate = '<br>';

	var checkBoxesRegExp = /\[( |x)\]/g;
	var checkBoxesTemplate = function (match, group1) {
		return '<input type="checkbox" disabled' + (group1.toLowerCase() === 'x' ? ' checked' : '') + '>'
	};


	/**
	 * markdown parser
	 * 
	 * @param  {string} markdown
	 * @return {string}
	 */
	function md (markdown) {
		if (!markdown) {
			return '';
		}

		var code = [];
		var index = 0;
		var length = markdown.length;

		// to allow matching trailing paragraphs
		if (markdown[length-1] !== '\n' && markdown[length-2] !== '\n') {
			markdown += '\n\n';
		}

		// format, removes tabs, leading and trailing spaces
		markdown = (
			markdown
				// collect code blocks and replace with placeholder
				// we do this to avoid code blocks matching the paragraph regexp
				.replace(blockCodeRegExp, function (match, lang, block) {
					var placeholder = '{code-block-'+index+'}';
					var regex = new RegExp('{code-block-'+index+'}', 'g');

					code[index++] = {lang: lang, block: block.replace(resc, unicode), regex: regex};

					return placeholder;
				})
				// XSS script tags
				.replace(XSSFilterRegExp, XSSFilterTemplate)
				// XSS image onerror
				.replace(XSSFilterImageRegExp, XSSFilterImageTemplate)
				// filter events
				.replace(eventsFilterRegExp, eventsFilterTemplate)
				// tabs
				.replace(removeTabsRegExp, '')
				// blockquotes
				.replace(blockQuotesRegExp, blockQuotesTemplate)
				// images
				.replace(imagesRegExp, imagesTemplate)
				// headings
				.replace(headingsRegExp, headingsTemplate)
				// headings h1 (commonmark)
				.replace(headingsCommonh1RegExp, headingsCommonh1Template)
				// headings h2 (commonmark)
				.replace(headingsCommonh2RegExp, headingsCommonh2Template)
				// horizontal rule 
				.replace(horizontalRegExp, horizontalTemplate)
				// checkboxes
				.replace(checkBoxesRegExp, checkBoxesTemplate)
				// filter html
				.replace(htmlFilterRegExp, htmlFilterTemplate)
				// filter css
				.replace(cssFilterRegExp, cssFilterTemplate)
				// paragraphs
				.replace(paragraphsRegExp, paragraphsTemplate)
				// inline code
				.replace(inlineCodeRegExp, inlineCodeTemplate)
				// links
				.replace(linksRegExp, linksTemplate)
				// unorderd lists
				.replace(listUlRegExp1, listUlTemplate).replace(listUlRegExp2, '')
				// ordered lists
				.replace(listOlRegExp1, listOlTemplate).replace(listOlRegExp2, '')
				// strong
				.replace(strongRegExp, strongTemplate)
				// emphasis
				.replace(emphasisRegExp, emphasisTemplate)
				// strike through
				.replace(strikeRegExp, strikeTemplate)
				// line breaks
				.replace(lineBreaksRegExp, lineBreaksTemplate)
				// filter inline js
				.replace(XSSFilterInlineJSRegExp, XSSFilterInlineJSTemplate)
		);

		// replace code block placeholders
		for (var i = 0; i < index; i++) {
			var item = code[i];
			var lang = item.lang;
			var block = item.block;

			markdown = markdown.replace(item.regex, function (match) {
				return '<pre><code class="language-'+lang+'">'+block+'</code></pre>';
			});
		}

		return markdown.trim();
	}

	return md;
}));